crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -481,6 +481,44 @@ def upgrade_all_compute_nodes
         true
       end
 
+      # Take a list of nodes and execute given script at each node in the background
+      # Wait until all scripts at all nodes correctly finish or until some error is detected
+      def execute_scripts_and_wait_for_finish(nodes, script, seconds)
+        nodes.each do |node|
+          ssh_status = node.ssh_cmd(script).first
+          if ssh_status != 200
+            save_error_state("Failed to connect to node #{node.name}!")
+            raise "Execution of script #{script} has failed on node #{node.name}."
+          end
+        end
+
+        scripts_status = {}
+        begin
+          Timeout.timeout(seconds) do
+            nodes.each do |node|
+              # wait until sript on this node finishes, than move to check next one
+              loop do
+                finished = node.script_finished(script)
```

Comment:
Nitpick: I'd suggest to rename this (both, the variable and the method) to ```script_status```. By just looking at the name, I expected ```script_finished``` to just return a boolean (which it doesn't)